src/test/java/jenkins/branch/matchers/Extracting.java

```diff
@@ -0,0 +1,56 @@
+package jenkins.branch.matchers;
+
+import org.hamcrest.Description;
+import org.hamcrest.Matcher;
+import org.hamcrest.TypeSafeDiagnosingMatcher;
+
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+public class Extracting<Element, Property> extends TypeSafeDiagnosingMatcher<Iterable<Element>> {
+
+    private final Function<? super Element, Property> extractor;
+    private final Matcher<?> nextMatcher;
+
+    private Extracting(Function<? super Element, Property> extractor, Matcher<?> nextMatcher) {
+        this.extractor = extractor;
+        this.nextMatcher = nextMatcher;
+    }
+
+    @Override
+    protected boolean matchesSafely(Iterable<Element> inColl, Description mismatchDescription) {
+        final Iterable<Property> collection = StreamSupport.stream(inColl.spliterator(), false)
+                .map(extractor)
+                .collect(Collectors.toList());
+
+        final boolean matches = nextMatcher.matches(collection);
+
+        if (!matches) {
+            nextMatcher.describeMismatch(collection, mismatchDescription);
+        }
+        return matches;
+    }
+
+
+    @Override
+    public void describeTo(Description description) {
+        description
+                .appendDescriptionOf(nextMatcher);
+    }
+
+    /**
+     * Extracts the specified property from a collection of things, and then lets you assert on the extracted properties. (The analog of AssertJ's <code>assertThat(things).extracting("foo")</code>.)
+     * <p>
+     * Use this in a matcher chain. For example, if you have list of POJOs with a property 'foo', you could write <code>assertThat(list, extracting(t -> t.getFoo(), hasItems("a", "b")))</code>.
```

Comment:
Is this something you wrote for this PR? Would suffice just to

```java
assertThat(projectFactories.stream().map(MultiBranchProjectFactory::getDescriptor).toArray(), arrayHasItem(ExtensionList.lookupSingleton(ConfigRoundTripDescriptor.class)));
```

or something along those lines.

Reply:
I did, since I wanted something equivalent to AssertJ's `extracting` matcher but couldn't use AssertJ here. Thinking I could also send this upstream to Hamcrest later if they're interested.